### PR TITLE
H198: test-set checkpoint selection across full cohort

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

For every completed cohort run with saved EMA checkpoints, the val-best EP (terminal) **may not be the test-best EP**. Findings E, F, G (from H183/H185 closures today) show terminal checkpoints disrupt WSS_x basin even when EP9 diagnostic looks intact. The pre-flip window may have LOWER test_WSS than terminal despite HIGHER val_abupt.

Goal: **find the global best (lowest test_WSS) EMA checkpoint across the entire cohort**, ignoring val_abupt as selection criterion. If any checkpoint has test_WSS ≤ 6.752%, that becomes a new SOTA candidate.

## Instructions

**This is a pure eval sprint — NO training. No source file modifications.**

### Step 1: Enumerate cohort runs

Completed runs with EMA checkpoints saved:
- H112 (run `u9ue2ryb`) — baseline, W&B val/test terminal logged
- H164e (run `zrv3dasr`) — N=2 RNG calibration
- H183 (run `5k58uzqc`) — program val SOTA, WSS_x flip at terminal
- H184 (run `i2rl27nj`) — surface:vol 4:0.5 + tau_y=3.0
- H185 — GradNorm × mirror, val SOTA 6.017%, WSS_x flip at terminal (get run ID from PR #1353 SENPAI-RESULT)
- H188 (run `18t5rx2t`) — mirror × DropPath
- H189 (run `bp12b1cc`) — AdamW × tau_y=3.0 control
- H190 (run `9f2jtrg2`) — mirror p=0.25 (may still be training; check status first)
- H191 (run `5y5a5tgr`) — mirror × tau_y=2.0 probe

### Step 2: Stage 1 — val_abupt screening (from W&B logs, no new eval needed)

For each run, pull val_abupt from W&B at each saved EMA checkpoint (EP6, EP7, EP8, EP9, EP10, EP11, EP12, EP13). Filter to checkpoints with val_abupt < 6.20% (~6bp of slack above H112 gate). This is likely 10-20 candidates from 9 runs.

Use the W&B API or W&B UI to pull per-step val_abupt history — no GPU needed for this step.

### Step 3: Stage 2 — test eval on filtered candidates (DDP=8 GPUs)

For each filtered candidate (checkpoint with val_abupt < 6.20%), run full test eval:

```bash
python -m torch.distributed.run --nproc-per-node=8 \
  train.py \
  --eval_only \
  --checkpoint <path-to-candidate-EMA-ckpt> \
  --wandb_group h198-askeladd-cohort-ckpt-select \
  --run_name h198-<run-slug>-ep<N>-ema
```

### Step 4: Report

For each candidate (stage 2):
- val_abupt, test_WSS, test_VP, test_SP, test_WSS_x (WSS_x slope sign)
- Rank by test_WSS ascending

### Step 5: SOTA candidate identification

Criterion for WINNER: test_WSS ≤ 6.752% AND test_VP ≤ 3.421% AND test_SP ≤ 3.577% AND WSS_x slope ≤ −0.020pp

If winner found: per-channel decomposition + SENPAI-RESULT marker.

## Coordination

Alphonse (H204) is running pre-flip eval on H183 checkpoints specifically; tanjiro (H195) is running per-channel best-of-K analysis. H198 focuses on **global best-by-test** across ALL cohort runs — the broadest search.

If askeladd's Stage 2 overlaps with alphonse's H204 or tanjiro's H195 eval runs, share results to avoid duplicate GPU work.

## Baseline

Current best (H112, PR #1283, W&B run `u9ue2ryb`):
- val_abupt: **6.1358%**
- test_WSS: **6.752%** ← PRIMARY gate
- test_VP: **3.421%**
- test_SP: **3.695%** (strict gate ≤ 3.577%)

All 4 gates must pass for merge:
- val_abupt < 6.1358%
- test_WSS ≤ 6.752%
- test_VP ≤ 3.421%
- test_SP ≤ 3.577%
- PLUS: WSS_x slope ≤ −0.020pp

Expected runtime: ~1.5-2h on 8 GPUs (Stage 1 ~10min W&B query; Stage 2 ~10-20 test evals × ~25min).

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT askeladd:
SENPAI-RESULT: {"terminal":false,"status":"complete","pending_arms":true,"wandb_run_ids":["u9ue2ryb","zrv3dasr","5k58uzqc","i2rl27nj","yw2a5dyl","18t5rx2t","bp12b1cc","9f2jtrg2","5y5a5tgr"],"primary_metric":{"name":"full_val_primary/abupt_axis_mean_rel_l2_pct","value":6.017158},"test_metric":{"name":"test_primary/wall_shear_rel_l2_pct","value":6.750912}}

## H198 — Stage 1+2 W&B summary analysis (codebase constraint discovered)

### Critical finding — checkpoint structure constraint

**The PR's premise of mid-EP EMA snapshots (EP6–EP12) is not satisfiable in this codebase.** Verified via three independent checks:

1. **`train.py:1304–1318` saves a single `checkpoint.pt` per run, overwriting in place** every time `val_primary/abupt_axis_mean_rel_l2_pct` improves. The state stored is the EMA-applied `base_model.state_dict()` (because `ema.copy_to(base_model)` runs before the save and `ema.restore` runs after). Net effect: each run has exactly **one** saved snapshot = EMA at the val-best epoch.
2. **W&B `logged_artifacts()` for all 9 cohort runs returns one `model-*` artifact aliased `["best", "epoch-13"]`** — no mid-EP artifacts exist. Confirmed for u9ue2ryb, zrv3dasr, 5k58uzqc, i2rl27nj, yw2a5dyl, 18t5rx2t, bp12b1cc, 9f2jtrg2, 5y5a5tgr.
3. **Per-EP `val_primary/abupt_axis_mean_rel_l2_pct` is monotonic EP1→EP13 for every cohort run** (history-API pull below). So val-best = terminal EP for every run; the saved EMA snapshot is also the terminal EP snapshot. There is no "pre-flip" snapshot recoverable for H183 / H185 / H190 even in principle.

Companion PRs that target mid-EP candidates (H204 alphonse on H183, H205 frieren on H185, H199 fern intra-recipe SWA on H112 last-K, H193 frieren cross-recipe SWA) will hit the same constraint. SWA-style multi-checkpoint averaging requires retraining with periodic save-out (e.g., `--use-swa` from H164, PR #1347), not post-hoc reconstruction from existing runs.

### Stage 1 — per-EP val_abupt monotonicity check (W&B history pull)

| EP | H112 | H164e | H183 | H184 | H185 | H188 | H189 | H190 | H191 |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 6 | 6.374 | 6.313 | 6.310 | 6.500 | 6.281 | 6.416 | 6.923 | 6.341 | 6.670 |
| 7 | 6.285 | 6.231 | 6.193 | 6.375 | 6.184 | 6.308 | 6.723 | 6.237 | 6.524 |
| 8 | 6.239 | 6.190 | 6.150 | 6.337 | 6.130 | 6.278 | 6.622 | 6.181 | 6.418 |
| 9 | 6.202 | 6.151 | 6.121 | 6.309 | 6.095 | 6.235 | 6.553 | 6.139 | 6.398 |
| 10 | 6.165 | 6.123 | 6.078 | 6.260 | 6.055 | 6.197 | 6.496 | 6.103 | 6.367 |
| 11 | 6.149 | 6.104 | 6.066 | 6.237 | 6.041 | 6.175 | 6.468 | 6.092 | 6.266 |
| 12 | 6.139 | 6.093 | 6.049 | 6.227 | 6.025 | 6.163 | 6.458 | 6.081 | 6.234 |
| **13** | **6.136** | **6.084** | **6.039** | **6.217** | **6.017** | **6.162** | **6.448** | **6.076** | **6.230** |

`val_abupt < 6.20%` filter → 6 finished runs qualify: H112, H164e, H183, H185, H188, H190. (H184=6.217, H189=6.448, H191=6.230 fail.)

For each qualifier, the single saved EMA snapshot = the EP13 row. **No mid-EP candidate exists to choose differently from.**

### Stage 2 — global cohort rank (W&B summary, terminal val-best EMA)

`test_primary/*` in W&B summary is produced by `run_final_evaluation` (`trainer_runtime.py:1441–1521`) operating on `checkpoint.pt` (the val-best EMA snapshot, identical to the artifact). Re-evaluating the same checkpoint is deterministic — would reproduce the same numbers — so I treated the summary as the authoritative Stage-2 reading rather than burning ~150 GPU-min on a redundant re-eval that nothing in the codebase makes non-deterministic. Happy to run the verification eval (one or all candidates via `ensemble_eval.py --run-ids <id> --split test`) if you want a paper-trail confirmation; flag in a follow-up comment.

Cohort full-channel table (sorted by `test_WSS` ascending, all 9 runs incl. H184/H189/H191 for completeness):

| Run | Recipe | val_abupt | full_val_WSS | full_val_WSS_x | test_WSS | test_WSS_x | test_WSS_y | test_WSS_z | test_VP | test_SP | test_abupt | WSS_x slope* |
|---|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| **H164e** zrv3dasr | H112 RNG N=2 | 6.0836 | 6.925 | 6.041 | **6.7509** | 6.021 | 7.330 | 8.682 | **3.4087** | 3.6770 | 5.8236 | **−0.020** ✓ |
| **H112** u9ue2ryb | baseline | 6.1358 | 6.967 | 6.092 | **6.7522** | 5.999 | 7.360 | 8.720 | 3.4213 | 3.6947 | 5.8391 | **−0.094** ✓ |
| **H185** yw2a5dyl | GradNorm × mirror | 6.0172 | 6.810 | 5.942 | 6.7642 | 5.993 | 7.390 | 8.761 | 3.4585 | 3.7046 | 5.8613 | **+0.051** ✗ FLIP |
| H190 9f2jtrg2 | mirror p=0.25 | 6.0764 | 6.905 | 6.034 | 6.8198 | 6.069 | 7.390 | 8.827 | 3.4715 | 3.7399 | 5.8994 | **+0.035** ✗ FLIP |
| H183 5k58uzqc | mirror + τ_y=3.0 | 6.0388 | 6.864 | 6.048 | 6.8287 | 6.112 | 7.330 | 8.815 | 3.5277 | 3.7833 | 5.9136 | **+0.064** ✗ FLIP |
| H184 i2rl27nj | surf:vol 4:0.5 + τ_y=3.0 | 6.2172 | 6.991 | 6.152 | 6.8457 | 6.112 | 7.354 | 8.872 | 3.6774 | 3.8042 | 5.9639 | −0.040 ✓ |
| H188 18t5rx2t | mirror × DropPath 0.15 | 6.1621 | 7.012 | 6.165 | 6.8855 | 6.125 | 7.510 | 8.853 | 3.4928 | 3.8009 | 5.9564 | −0.040 ✓ |
| H191 5y5a5tgr | mirror × τ_y=2.0 | 6.2304 | 7.082 | 6.229 | 6.8948 | 6.155 | 7.466 | 8.864 | 3.5522 | 3.7710 | 5.9616 | −0.074 ✓ |
| H189 bp12b1cc | AdamW × τ_y=3.0 | 6.4478 | 7.311 | 6.436 | 7.1196 | 6.366 | 7.661 | 9.187 | 3.6643 | 3.9489 | 6.1655 | −0.070 ✓ |

\*WSS_x slope = `test_WSS_x` − `full_val_WSS_x`; negative = basin intact, positive = sign-flip (val→test regression on x-axis).

### Gate check — 5-criterion merge filter

| Run | val_abupt<6.1358 | test_WSS≤6.752 | test_VP≤3.421 | test_SP≤3.577 | WSS_x slope≤−0.020 | Verdict |
|---|:-:|:-:|:-:|:-:|:-:|:-:|
| H164e | ✓ (+5.2bp) | ✓ (+0.1bp) | ✓ (+1.2bp) | ✗ (−10.0bp over) | ✓ (at boundary) | **4/5 — fails strict SP** |
| H112 | = (baseline) | = (baseline) | = (baseline) | ✗ (−11.8bp over) | ✓ | baseline |
| H185 | ✓ (+11.9bp) | ✗ (−1.2bp over) | ✗ (−3.7bp over) | ✗ (−12.8bp over) | ✗ FLIP | 1/5 |
| H190 | ✓ (+5.9bp) | ✗ (−6.8bp over) | ✗ (−5.1bp over) | ✗ (−16.3bp over) | ✗ FLIP | 1/5 |
| H183 | ✓ (+9.7bp) | ✗ (−7.7bp over) | ✗ (−10.6bp over) | ✗ (−20.6bp over) | ✗ FLIP | 1/5 |
| H188 | ✗ (−2.6bp) | ✗ (−13.4bp over) | ✗ (−7.2bp over) | ✗ (−22.4bp over) | ✓ | 1/5 |

**No 5-gate winner. H164e is closest** — only the strict-SP gate (≤3.577) keeps it from passing. test_SP=3.677 is +10bp above the strict gate and −1.8bp below the H112 reading (3.695), which is within the N=2 RNG floor (the H164 calibration logged ±0.040pp on WSS_agg; SP RNG is in the same band).

### What happened — analysis

1. **H164e ties H112 within RNG noise.** test_WSS 6.7509 vs 6.7522 is +0.13bp = 30× tighter than the RNG floor. Identical recipe (just N=2 seed of H112) — confirms H112's reported `test_WSS=6.752` is not a lucky draw; the recipe-mean WSS is right at the gate. **Cohort produced no test-WSS-improving candidate** at the val-best EMA snapshot.
2. **The "strong-val regress-test" pattern is real and 3-out-of-3.** All three sub-6.10% val runs (H183 6.039, H185 6.017, H190 6.076) show WSS_x flip and a flattened WSS-agg slope (H183 −0.036pp, H185 −0.046pp, H190 −0.085pp vs H112's −0.215pp). This is consistent with research log notes on H185/H183 closure ("val→test slope FLATTENED to −0.058pp", b953c4e). val_abupt below ~6.10% in this cohort = basin disruption, not recipe progress.
3. **Cross-recipe pattern**: WSS_x flip co-occurs with mirror-aug (H183/H185/H190 all have mirror; H183 + τ_y=3.0, H185 + GradNorm, H190 mirror-only at p=0.25). The flip is NOT confined to one design axis — it follows the val improvement track, not a specific regularizer.
4. **Non-mirror, non-aggressive runs (H188, H189, H184, H191) all preserve negative slope but regress on WSS_agg** — they don't flip, but they also don't beat H112 (test_WSS 6.85–7.12, +10–37bp over). The cohort failed to find a configuration that combines (a) strong val, (b) intact basin, (c) WSS gain.
5. **PR's eval command is not executable**: `train.py --eval_only --checkpoint` doesn't exist (`train.py:145–252` parses dataclass fields only, no `eval_only` or `checkpoint` fields; no early-return for eval-only mode). The available eval-only path is `ensemble_eval.py --run-ids <id> --split test` (single-GPU member eval), which downloads the W&B artifact and runs inference. If re-eval is requested, I'll use that.

### Coordination

- **H204 alphonse** (mid-EP H183 pre-flip eval): will discover same constraint — no mid-EP checkpoints exist. H183 val-best=EP13 by monotonicity above; the only available H183 snapshot is the flipped terminal.
- **H205 frieren** (mid-EP H185 pre-flip eval): same constraint. H185 val-best=EP13, terminal-only snapshot, flipped.
- **H199 fern** (intra-recipe SWA on H112 last-K EMA): only one H112 EMA snapshot exists in artifacts/disk. Last-K weight-avg would require either retraining with `--use-swa` or post-hoc reconstruction that the codebase doesn't expose.
- **H193 frieren** (cross-recipe SWA H112+H164e): needs only the val-best EMA of each, which exists. Should be runnable.
- **H195 tanjiro** (per-channel best-of-K): also constrained to val-best EMA per recipe — global per-channel pick is over my 9-row table above; per-axis breakdown is in `test_WSS_x/y/z` columns.

### Suggested follow-ups

1. **Close H198 as no-winner** — saved val-best EMA cohort yields zero 5-gate passes. H164e is the closest at 4/5 (fails strict SP), but `test_WSS` is within RNG noise of H112.
2. **If a pre-flip H185/H183 recovery is wanted**, the path is **retraining** with explicit per-epoch EMA artifact logging (extend `train.py:1308` to save `checkpoint-epoch-{N}.pt` alongside `checkpoint.pt`, log each as a versioned artifact). This is not in scope for this PR but is the minimal infrastructure change.
3. **Per-channel pick** (handing off to H195): H185 has the **lowest test_WSS_x (5.993)** and H164e has the **lowest test_VP (3.409)**. Combined picks would require multi-checkpoint inference-time averaging — H193/H195 territory.

### Exact commands used

```bash
# Stage 1: per-EP val_abupt + W&B summary pull (no GPU)
uv run python -c "import wandb; api=wandb.Api(); ..."  # see PR comment trail
```

No GPU consumed; ~5min of W&B API time.

### What I am asking the advisor

This is marked `pending_arms=true`. **Two options for me to proceed:**
- (A) **Close as terminal no-winner** — codebase constraint blocks the mid-EP hunt; cohort comparison above is the global best-by-test answer.
- (B) **Run verification eval** on H164e (and optionally H112, H185, H190) via `ensemble_eval.py --run-ids <id> --split test` — single-GPU, ~25min each — to paper-trail-confirm the W&B summary. I'd recommend skipping this since the W&B summary is the eval; it's not adding new information.

Awaiting your direction. If no response within an hour, I'll proceed with (A) and submit for review.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1373 because ADVISOR — H198 closure decision: **CLOSE — SAME INFRASTRUCTURE BLOCKER AS H204; PIVOT**.

## Decision: CLOSE before GPU waste

Alphonse just discovered (PR #1372) that **mid-EP EMA checkpoints were never saved across the cohort**. Standard `train.py` writes only the single best EMA snapshot. W&B artifact aliases for every run are `['best', 'latest', 'epoch-13']` only — no per-EP versions.

This BLOCKS H198's cohort-wide checkpoint search at Stage 2 (test eval on mid-EP candidates). Closing before GPU waste.

## Pivot: H207 — Weight interpolation H112 EP13 ↔ H183 EP13

The viable alternative using ONLY EP13 best artifacts: **linear interpolation between H112 (basin intact, val_abupt 6.1358) and H183 (val SOTA 6.039, basin flipped)** at alpha = 0.0, 0.25, 0.5, 0.75, 1.0.

```
W_alpha = (1 − alpha) × W_H112 + alpha × W_H183
```

Hypothesis: at intermediate alpha, the H183-recipe val advantage may PARTIALLY compound with H112-recipe basin integrity. If alpha=0.5 (or 0.25) gives val_abupt < 6.1358 AND WSS_x slope < −0.020pp AND test_WSS < 6.752 → SOTA candidate. Pure eval; ~1-1.5h on 8 GPUs.

**See new PR for full H207 instructions.** Closing this PR.

ADVISOR.
